docs/design/2018-07-22-enhance-propagations.md

```diff
@@ -0,0 +1,189 @@
+|     └─TableScan_11      | 2.00  | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo             |
++-------------------------+-------+------+-------------------------------------------------------------------------+
+7 rows in set (0.00 sec)
+```
+
+## Rationale
+
+Constraint propagation is commonly used as logical plan optimization in traditional databases. For example, [this doc](https://dev.mysql.com/doc/internals/en/optimizer-constant-propagation.html) explains some details of constant propagations in MySQL. It is is also widely adopted in distributed analytical engines, like Apache Hive, Apache SparkSQL, Apache Impala, et al. Those engines usually query on a huge amount of data.
+
+### Advantages:
+
+ Constraint propagation brings more detailed, explicit constraints to each data source involved in a query. With those constraints, we can filter data as early as possible, and thus reduce disc/network I/O and computational overheads during the execution of a query. In TiDB, most propagated filters can be pushed down to the storage level (TiKV), as a Coprocessor task, and leads to the following benefits:
+
+ * Apply the filters at each storage segment (Region), which make the calculation distributed.
+
+ * When loading data, skip some partitions of a table if the partitioning expression doesn't pass the filter.
+
+ * For the columnar storage format to be supported in the future, we may apply some filters directly when accessing the raw storage.
+
+ * Reduce the data transfered from TiKV to TiDB.
+
+### Disadvantages:
+
+ Constraint propagation may brings unnecessary filters and leads to unnecessary overheads during a query, this is mostly due to the fact that logical optimization doesn't take data statistics into account, for example:
+
+ For a query `select * from t0, t1 on t0.a = t1.a where t1.a < 5`, we get a propagation `t0.a < 5`, but if all `t0.a` is greater than 5, applying the filter brings unnecessary overheads.
+
+Considering the trade-off, we still gain a lot of benefits from constraint propagation in most of cases; hence it still can be treated it as useful.
+
+## Compatibility
+
+All rules mentioned in this proposal are logical plan optimization, which do not change the semantics of a query, and thus this proposal will not lead to any compatibility issue
+
+## Implementation
+
+Here are rough ideas about possible implementations:
+
+ * For proposal #1, we can extend the current `propagateConstantSolver` to support wider types of operators from column equality.
+
+ * For proposal #2, `propagateConstantSolver` is also an applicable way to add `NotNULL` filter(`not(isnull())`), but should examine whether the column has the `NotNULL` constraint already.
+
+ * For proposal #3, the [ranger](https://github.com/pingcap/tidb/blob/6fb1a637fbfd41b2004048d8cb995de6442085e2/util/ranger/ranger.go#L14) may be useful to help us collect and fold comparison constraints
```

Comment:
currently, ranger only comes in when we are computing ranges for index scan, or for table scan with filter on RowID. If we want to impose general expression simplification, we need to extract infrastructure functionalities in `util/ranger/points.go` into a general module.